design/open/spark-outputcommitter/committer.md

```diff
@@ -0,0 +1,148 @@
+completion records are processed and the objects magically appear.
+
+The [staging committer][staging] (aka "Netflix committer") is an older
+committer for S3A.  It requires a "staging" filesystem on HDFS and offers
+manual modes for overwriting outputs ("conflict resolution").
+
+These issues remain with the magic and staging committer:
+
+* Not atomic.  Multipart objects appear sequentially.  Finding a "_success"
+  object indicator is still required before a directory tree can be
+  processed.
+* The magic committer requires "magic committer" support in the FileSystem,
+  in order to write `__magic` objects to a _different_ path.  (Current
+  lakeFSFS does not contain this support.)
+* The magic committer is somewhat new:
+  > It’s also not been field tested to the extent of Netflix’s committer;
+  > consider it the least mature of the committers.
+* Documentation is still somewhat lacking.
+
+Other OutputCommitters are not as good, and _rename by copying_.  This is
+both slow and highly non-atomic, as well as making it difficult to recover
+from failed attempts.  The current recommendations are to use the
+partitioned staging committer for overwriting or updating partitioned data
+trees, and the magic committer is most other cases.
+
+## LakeFSOutputCommitter
+
+We propose to leverage the atomic capabilities of lakeFS to write a specific
+OutputCommitter for lakeFSFS.  In the initial version, it will branch out,
+prepare the desired output, and merge back in to commit.  Aborting will be
+done by dropping the branch (or repurposing it if the same job ID is
+requested again).  Regular lakeFS retention can handle dropping data
+objects; we might want to add file patterns to retention to allow temporary
+objects to be dropped rapidly.
+
+# How
+
+## Sample flow
+
+* User configures lakeFSFS and configures LakeFSOutputCommitter as the
+  OutputCommitter for `lakefs` protocol Paths.  **Possibly** lakeFSFS will
+  set up this OutputCommitter by default.
+* **Setup (job/task TBD)**: Create a new branch for this job/task.  Its name
+  is predictable from the job ID and/or task ID, so can be easily found
+  again.  Immediately delete the entire subree of the intended output
+  path, and possibly delete branches of previous tasks.
+* **Write objects**: Everything is written to its correct path on the branch.
+* **Commit**: Merge back to the original branch.
+
+## Properties
+
+* The merge is performed by lakeFS so it is **atomic**.
+* In the **single-writer case** the merge succeeds: no other operations occur on the subtree.
+* **In-place updates** work: the old objects are deleted and replaced by new
+  objects.  This is true regardess of partitioning etc.
+* **Multiple writers** are detected: the first to commit succeeds, the
+  others fail.
+* (Conflicting) **non-OutputCommitter writes are detected** and clearly
+  handled.  LakeFSOutputCommitter cab achieve its correct semantics
+  regardless of other writers used.
```

Comment:
This one needs more clarification. Why is this possible?

Does this depend on the fact that lakeFS fails to merge when the destination branch is "dirty"?

Reply:
You're correct (of course; as always), so I added some text to say this.